### PR TITLE
Public key quadratic scheme

### PR DESCRIPTION
**Pull request by tilenmarc**
This PR implements a FE public key quadratic scheme for multivariate polynomials. The scheme allows to encrypt vectors `x` and `y` using a public key and to derive a FE key corresponding to a matrix `F` so that using the FE key one can decrypt the value `x^TFy` without revealing `x` or `y`.

Additionally, a partially function hiding inner product FE scheme (partFHIPE) is implemented on which the quadratic scheme is based.  partFHIPE is a public key FE scheme that allows to encrypt vectors and produce FE keys to be able to decrypt only the inner product of the encryption and chosen vector without revealing the encrypted or FE key vector. Public key encryption allows to encrypt only vectors from a chosen subspace. This way a functional encryption key does not reveal its corresponding inner product vector.

### CODE REVIEW

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- innerprod/fullysec/part_fh_ipe_test.go  15
- quadratic/quad.go  13
- innerprod/fullysec/part_fh_ipe.go  8
- quadratic/quad_test.go  10
         

Clones added
============
- data/vector.go  2
- data/matrix.go  2
- innerprod/fullysec/part_fh_ipe.go  2
- data/matrix_test.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/tilenmarc/gofe_2/pullRequest?prid=5056237&bid=16643720)

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- quadratic/quad_test.go  10
- innerprod/fullysec/part_fh_ipe.go  8
- innerprod/fullysec/part_fh_ipe_test.go  15
- quadratic/quad.go  13
         

Clones added
============
- data/matrix_test.go  2
- innerprod/fullysec/part_fh_ipe.go  2
- data/vector.go  2
- data/matrix.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/tilenmarc/gofe_2/pullRequest?prid=5056237&bid=16643720)

File: innerprod/fullysec/part_fh_ipe.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+	key := append(key1, key2...)
+
+	return key.MulG2(), nil
+}
+
+// Encrypt on input vector t encrypts vector x = Mt with the provided public key
+// (matrix M is specified in the public key). It returns a ciphertext vector.
+// Entries of Mt should not be greater then bound.
+// If encryption an failed, an error is returned.
```
Comment (miha-stopar) [inline innerprod/fullysec/part_fh_ipe.go]: 'an failed' - typo

File: innerprod/fullysec/part_fh_ipe_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		te.Fatalf("Error during random vector generation: %v", err)
+	}
+
+	// encrypt the vector
+	ciphertextMt, err := encryptor.Encrypt(t, pubKey)
+	if err != nil {
+		te.Fatalf("Error during encryption: %v", err)
+	}
+
+	// owner of the secret key key encrypt an arbitrary vector
```
Comment (miha-stopar) [inline innerprod/fullysec/part_fh_ipe_test.go]: 'key key encrypt' -> 'key encrypts'

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- quadratic/quad.go  13
- innerprod/fullysec/part_fh_ipe.go  8
- innerprod/fullysec/part_fh_ipe_test.go  15
- quadratic/quad_test.go  10
         

Clones added
============
- data/matrix_test.go  2
- data/matrix.go  2
- innerprod/fullysec/part_fh_ipe.go  2
- data/vector.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/tilenmarc/gofe_2/pullRequest?prid=5056237&bid=16643720)

**Pull request merged**